python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.types.v2;
+
+import java.util.ArrayDeque;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.function.Consumer;
+import org.sonar.plugins.python.api.LocationInFile;
+import org.sonar.python.semantic.v2.SymbolsModuleTypeProvider;
+
+public class LazyType implements PythonType {
+
+  String fullyQualifiedName;
+  private final Queue<Consumer<PythonType>> consumers;
+  private final SymbolsModuleTypeProvider symbolsModuleTypeProvider;
```

Comment:
I think it is better to operate with ProjectLevelTypesTable cause once you trigger lazy type resolution - it's type could be already resolved

Reply:
I'm still not sure on that one, because `LazyType` themselves are created in `SymbolsModuleTypeProvider` during some actual conversions from symbols to types.

While the current architecture is already a bit odd, I feel it would be even weirder if `SymbolsModuleTypeProvider` held a reference to `ProjectLevelTypeTable` simply for the sake of passing it to `LazyType`. I don't really have a strong opinion here though; and.I feel this is temporary code anyway since it will only exist while we are doing symbol to type conversions.